### ISSUE TITLE
Semantics for optional number parameters feel wrong

### ISSUE THREAD

**Issue by jenshalm** [finagle/finch#157]
I just stumbled across the implementation for optional number values. I think the behaviour is almost never what a user would expect. If an API has optional parameters, the usual semantics are that they can be left out or otherwise, if present, must be valid. Treating a missing optional value the same as an invalid value can easily lead to silent failure and unexpected behaviour. I would expect:

```
Some(int)       // if the param is present and valid
None            // if the param is missing
ValidationError // if the param is present but invalid
```

I can do a PR should there be agreement on this.

**Comment by jenshalm** [comment]
I'll add a concrete example to illustrate my point. Let's assume this route:

```
/london/food-shops
```

would by default return 10,000 rows, but I'm on a slow mobile connection, so would prefer to use the optional `limit` parameter, but thanks to chubby fingers I'll type:

```
/london/food-shops?limit=f100
```

With Finch the parameter would just be ignored, the same as if it would not be there and I would be waiting for the 10,000 rows (and probably not know why). Clearly I would expect to get a 400 here.

**Comment by benjumanji** [comment]
I'd agree that if a validation rule was specified I would expect it to enforced when a parameter is provided.

**Comment by vkostyukov** [comment]
Sounds good to me. Can we plan this be done at 0.6.0 milestone? There are a bunch of big/important changes in this release 0.5.0, so we probably want to feature-freeze it. The last feature I'm planing to release is #26. What do you think folks?

**Comment by jenshalm** [comment]
Can you wait for my weekend PRs? I am in the process of preparing 5 of them. Should all be ready on Sunday evening, so you can have a look how many of them you want to discuss for 0.5.0 timeframe and how many you want to defer? It's not crazily important though, if we still allow a few API changes for 0.6.0 and do not wait for too long for the next release.

**Comment by benjumanji** [comment]
We can anything we want before 1.0, including becoming a full featured ASN.1 serialization library that has nothing to do with the finagle!

*maniacal laughter*

**Comment by vkostyukov** [comment]
Sure! 5 PRs sounds crazy :) I'm fine with API changes, I just want to do them incrementally. Please, go ahead with the PRs, we will review them and merge either in 0.5.0 or in 0.6.0.

**Comment by vkostyukov** [comment]
I usually do short cycles between releases, like 2-3 new features and 20-25 commits (around a month).

**Comment by jenshalm** [comment]
Ok, then you'll probably just pick one or two of them for 0.5.0. Some are straightforward and not really controversial, while others might stir some debate. I guess everything that needs a debate is for 0.6.0 then (which is fine).

**Comment by vkostyukov** [comment]
Cool, thanks! I will be really smooth. The 0.5.0 release should happen in a week (since there is some documentation changes that should be done).

**Comment by vkostyukov** [comment]
I've also start thinking that we have to change the semantic for required params a bit. Let's say we would allow having an empty string be a required param (for now it throws error). And if users actually don't accept empty strings `""` they may mix `BeEmpty` predefined validation rule from #164. Like this:

```scala
val name = RequiredParam("name") shouldNot BeEmpty
```

**Comment by vkostyukov** [comment]
Yeah. Right. 99% params should not be empty. If we had this way of explicitly allowing the empty strings for params, we might change the behavior for optional prams to return `None` be default.

**Comment by jenshalm** [comment]
Should we change the title on this issue? My original concern was that invalid optional values lead to a `None` instead of an error, the refactoring in PR 4 already solves this issue. So I'd suggest I'll add "fixes #157" to that commit and we create a new issue "Unify handling of empty strings".

**Comment by vkostyukov** [comment]
Sounds great! I will create a new issue.
